pep-0546.txt

```diff
@@ -0,0 +1,141 @@
+Version: $Revision$
+Last-Modified: $Date$
+Author: Victor Stinner <victor.stinner@gmail.com>,
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 30-May-2017
+
+
+Abstract
+========
+
+Backport ssl.MemoryBIO and ssl.SSLObject classes from Python 3 to Python
+2.7 to enhance the overall security of Python 2.7.
+
+
+Rationale
+=========
+
+While Python 2.7 is getting closer to its end-of-line (scheduled for
+2020), it is still used on production and the Python community is still
+responsible for its security. And to facilitate the future adoption of
+:pep:`543`, which will improve security for Python3 users.
+
+This PEP does NOT propose a general exception for backporting new
+features to Python 2.7 - every new feature proposed for backporting will
+still need to be justified independently. In particular, it will need to
+be explained why relying on an independently updated backport on the
+Python Package Index instead is not an acceptable solution.
+
+
+PEP 543
+-------
+
+The :pep:`543` defines a new TLS API for Python which would enhance the
+Python security: give access to the root certificate authorities on
+Windows and macOS by using native APIs, instead of OpenSSL. A side effect
+is that it gives access to certificates installed locally by system
+administrators, allowing to use "company certificates" without having to
+modify each Python application and so validate correctly TLS
+certificates (instead of having to ignore or bypass the TLS certificate
+validation).
+
+For practical reasons, Cory Benfield would like to first implement an
+I/O-less class similar to ssl.MemoryBIO and ssl.SSLObject for the
+:pep:`543`, and provide a second class based on the first one to use
+sockets or file descriptors.  This design would help to structure the code
+to support more backends and simplify testing and auditing. Later,
+optimized classes using directly sockets or file descriptors may be
+added for performance.
+
+While the :pep:`543` defines an API, the PEP would only make sense if it
+comes with at least one complete and good implementation. The first
+implementation will be based on the ``ssl`` module of the Python
+standard library.
+
+In a perfect world, all applications would already run on Python 3 since
+Python 3.0 was released. In practice, many applications still run on
+production on top of Python 2.7. To make the new TLS API more widely
+used, it should be usable on all Python versions currently supported:
+Python 2.7, 3.5, 3.6. Otherwise, some applications would have to wait
+until they drop Python 2 support to be able to use the new TLS API.
+
+Delaying adoption of the PEP 543 API means delaying the adoption for
+security improvements for Python 3 users as well.
+
+
+requests, pip and ensurepip
+---------------------------
+
+There are plans afoot to look at moving Requests to a more event-loop-y
+model, and doing so basically mandates a MemoryBIO. In the absence of a
```

Comment:
For what it's worth, Twisted used to implement TLS via the wrap_socket-esque approach (since this is what OpenSSL strongly encourages you to do).  Despite extensive test coverage and plenty of real-world usage, it never *really* worked right, and when we finally managed to switch over to the in-memory BIO model, *dozens* of bugs were fixed overnight, the whole system got considerably more reliable.

These edge cases manifest most significantly when using embedded systems, low-spec hardware (think raspberry pi), or weird operating systems.  I do still occasionally experience weird flakiness when using Tornado event loops on this kind of hardware that doesn't happen with Twisted, and tellingly, doesn't happen with Twisted's TLS support using `tornado.platform.twisted`.  I can't prove that there are bugs (the bugs tend to be extremely squirrely race-conditions that are hard to reproduce in-vitro) but I would definitely suspect there are some there.